### PR TITLE
Log info before asserts

### PR DESCRIPTION
**Pull request by ashtul**
Add macro with `RedisModule_Log` before `assert` crashes server.

@MeirShpilraien @mnunberg @rafie please review macro.
Used the macro in a few places as an example.

### CODE REVIEW

**Comment by ashtul** [comment]
The coverage test fails. Any idea why? It is on `test_ext.py` file.

https://app.circleci.com/jobs/github/RediSearch/RediSearch/6950

File: src/rm_assert.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#else
+// `ctx` can be NULL
+#define RS_LOG_ASSERT_FMT(ctx, condition, fmt, ...)                                \
+    if (!(condition)) {                                                            \
+        RedisModule_Log(ctx, "warning", "File %s, function %s, Line %d - "      \
+                fmt, __FILE__, __func__,  __func__, __LINE__, __VA_ARGS__);     \
+        assert(#condition/* ## "failed"*/);                                     \
+    } 
+
+// `ctx` can be NULL
```
Comment (MeirShpilraien) [inline src/rm_assert.h]: ctx can not be null, this should work on redis version 4 as well, create some static ctx and reuse it.

File: src/rm_assert.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+#define RS_LOG_ASSERT(ctx, condition, fmt, ...)    (__ASSERT_VOID_CAST (0))
+#define RS_LOG_ASSERT_STR(ctx, condition, str)     (__ASSERT_VOID_CAST (0))
+#else
+// `ctx` can be NULL
+#define RS_LOG_ASSERT_FMT(ctx, condition, fmt, ...)                                \
+    if (!(condition)) {                                                            \
+        RedisModule_Log(ctx, "warning", "File %s, function %s, Line %d - "      \
+                fmt, __FILE__, __func__,  __func__, __LINE__, __VA_ARGS__);     \
+        assert(#condition/* ## "failed"*/);                                     \
```
Comment (MeirShpilraien) [inline src/rm_assert.h]: This can be just assert false no? an even better option is to cause segfault so we will even get the full crash report.

**Comment by ashtul** [comment]
Defiantly more assert. Didn't want to run ahead before consensus on `macro`.

File: src/rm_assert.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+#define RS_LOG_ASSERT_STR(ctx, condition, str)     (__ASSERT_VOID_CAST (0))
+
+#else
+
+#define RS_LOG_ASSERT_FMT(condition, fmt, ...)                                      \
+    if (!(condition)) {                                                             \
+        RedisModuleCtx* assertCtx = RedisModule_GetThreadSafeContext(NULL);         \
+        RedisModule_Log(assertCtx, "warning", "File %s, Function %s, Line %d - "    \
+                fmt, __FILE__, __func__, __LINE__, __VA_ARGS__);                    \
+        *((char *)NULL) = 0;                                                        \
```
Comment (MeirShpilraien) [inline src/rm_assert.h]: Please add a comment and explain why we do it.

File: src/document.c [hunk 4]
```diff
@@ -179,7 +179,8 @@ RSAddDocumentCtx *NewAddDocumentCtx(IndexSpec *sp, Document *b, QueryError *stat
   aCtx->next = NULL;
   aCtx->specFlags = sp->flags;
   aCtx->indexer = sp->indexer;
-  assert(sp->indexer);
+  RS_LOG_ASSERT(sp->indexer, "No indexer");
+  //assert(sp->indexer);
```
Comment (MeirShpilraien) [inline src/document.c]: Remove all those commented asserts

**Comment by ashtul** [comment]
Failure
```
       Start  58: PY_test_ext

58: Test command: /root/project/src/pytest/runtests.sh "--module=/root/project/build-coverage/redisearch.so" "-t" "/root/project/src/pytest/test_ext.py"
58: Test timeout computed to be: 1500
58: ++ cd /root/project/src/pytest
58: ++ pwd -P
58: + SCRIPT_DIR=/root/project/src/pytest
58: + '[' -d /root/project/src/pytest/RLTest ']'
58: + '[' -n '' ']'
58: + PYTHONPATH=:/root/project/src/pytest
58: + export PYTHONPATH
58: + '[' -e rltest.config ']'
58: + ARGS='@rltest.config '
58: + '[' -n '' ']'
58: + exec python -m RLTest @rltest.config --module=/root/project/build-coverage/redisearch.so -t /root/project/src/pytest/test_ext.py
58: test_ext:testExt
58: 	[ERROR]
58: 	Unhandled exception: Cannot establish connection Redis<ConnectionPool<UnixDomainSocketConnection<path=/root/project/build-coverage/logs/f13af1a7e3624f60a3440c001c3231f2-master.sock,db=0>>>: Error 111 connecting to unix socket: /root/project/build-coverage/logs/f13af1a7e3624f60a3440c001c3231f2-master.sock. Connection refused.
58: Traceback (most recent call last):
58:   File "/usr/local/lib/python2.7/dist-packages/RLTest/__main__.py", line 496, in _runTest
58:     fn()
58:   File "/root/project/src/pytest/test_ext.py", line 34, in testExt
58:     env = Env(moduleArgs='EXTLOAD %s' % EXTPATH)
58:   File "/usr/local/lib/python2.7/dist-packages/RLTest/env.py", line 170, in __init__
58:     self.start()
58:   File "/usr/local/lib/python2.7/dist-packages/RLTest/env.py", line 238, in start
58:     self.envRunner.startEnv()
58:   File "/usr/local/lib/python2.7/dist-packages/RLTest/redis_std.py", line 193, in startEnv
58:     self.waitForRedisToStart(con)
58:   File "/usr/local/lib/python2.7/dist-packages/RLTest/redis_std.py", line 132, in waitForRedisToStart
58:     wait_for_conn(con, retries=1000 if self.debugger else 200)
58:   File "/usr/local/lib/python2.7/dist-packages/RLTest/utils.py", line 22, in wait_for_conn
58:     raise Exception('Cannot establish connection %s: %s' % (conn, err1))
58: Exception: Cannot establish connection Redis<ConnectionPool<UnixDomainSocketConnection<path=/root/project/build-coverage/logs/f13af1a7e3624f60a3440c001c3231f2-master.sock,db=0>>>: Error 111 connecting to unix socket: /root/project/build-coverage/logs/f13af1a7e3624f60a3440c001c3231f2-master.sock. Connection refused.
58: Test Took: 20 sec
58: Total Tests Run: 1, Total Tests Failed: 1, Total Tests Passed: 0
58: Failed Tests Summary:
58: 	test_ext:testExt
58: 		Cannot establish connection Redis<ConnectionPool<UnixDomainSocketConnection<path=/root/project/build-coverage/logs/f13af1a7e3624f60a3440c001c3231f2-master.sock,db=0>>>: Error 111 connecting to unix socket: /root/project/build-coverage/logs/f13af1a7e3624f60a3440c001c3231f2-master.sock. Connection refused.
 58/109 Test  #58: PY_test_ext ...................................***Failed   20.21 sec
```
@rafie - can you take a look?

File: src/rmutil/rm_assert.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+#define RS_LOG_ASSERT(ctx, condition, fmt, ...)    (__ASSERT_VOID_CAST (0))
+#define RS_LOG_ASSERT_STR(ctx, condition, str)     (__ASSERT_VOID_CAST (0))
+
+#else
+
+#define RS_LOG_ASSERT_FMT(condition, fmt, ...)                                      \
+    if (!(condition)) {                                                             \
+        RedisModuleCtx* assertCtx = RedisModule_GetThreadSafeContext(NULL);         \
+        RedisModule_Log(assertCtx, "warning", "File %s, Function %s, Line %d - "    \
```
Comment (mnunberg) [inline src/rmutil/rm_assert.h]: RedisModule_Log(ctx, "%s failed on %s:%d (%s)", #condition, __FILE__, __LINE__, __func__);

File: src/rmutil/rm_assert.h [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+#ifdef NDEBUG
+
+#define RS_LOG_ASSERT(ctx, condition, fmt, ...)    (__ASSERT_VOID_CAST (0))
+#define RS_LOG_ASSERT_STR(ctx, condition, str)     (__ASSERT_VOID_CAST (0))
+
+#else
+
+#define RS_LOG_ASSERT_FMT(condition, fmt, ...)                                      \
+    if (!(condition)) {                                                             \
+        RedisModuleCtx* assertCtx = RedisModule_GetThreadSafeContext(NULL);         \
```
Comment (mnunberg) [inline src/rmutil/rm_assert.h]: RSDummyContext

**Pull request merged**